cloudinit/sources/DataSourceAeza.py

```diff
@@ -0,0 +1,172 @@
+    raise RequestException(f"Failed to retrieve data from {url} after {retries + 1} attempts.")
+
+
+def read_metadata(url, timeout_seconds=2, max_wait_seconds=120, sec_between_retries=2, retries=30):
+    """
+    Read user data from URL or raise error.
+    """
+    response = read_url(
+        url,
+        timeout_seconds=timeout_seconds,
+        max_wait_seconds=max_wait_seconds,
+        sec_between_retries=sec_between_retries,
+        retries=retries,
+    )
+    if not response.ok:
+        raise RuntimeError("unable to read metadata at %s" % url)
+    return util.load_json(response.content.decode())
+
+
+def read_data(url, timeout_seconds=2, max_wait_seconds=120, sec_between_retries=2, retries=30):
+    """
+    Read data from URL.
+    """
+    try:
+        response = read_url(
+            url,
+            timeout_seconds=timeout_seconds,
+            max_wait_seconds=max_wait_seconds,
+            sec_between_retries=sec_between_retries,
+            retries=retries,
+        )
+        if not response.ok:
+            raise RuntimeError("unable to read data at %s" % url)
+        return response.content
+    except HTTPError as e:
+        if e.response.status_code == 404:
+            return None
+        raise e
+
+
+def format_url(url, system_uuid=None):
+    if system_uuid is None:
+        system_uuid = dmi.read_dmi_data("system-uuid")
+    return url.format(id=system_uuid)
+
+
+class DataSourceAeza(sources.DataSource):
+
+    dsname = "Aeza"
+
+    def __init__(self, sys_cfg, distro, paths, ud_proc=None):
+        super().__init__(sys_cfg, distro, paths, ud_proc)
+
+        self.ds_cfg = util.mergemanydict([self.ds_cfg, BUILTIN_DS_CONFIG])
+
+        url_params = self.get_url_params()
+        self.timeout_seconds = url_params.timeout_seconds
+        self.max_wait_seconds = url_params.max_wait_seconds
+        self.retries = url_params.num_retries
+        self.sec_between_retries = url_params.sec_between_retries
+
+        self.metadata_address = self.ds_cfg["metadata_url"]
+        self.userdata_address = self.ds_cfg["userdata_url"]
+        self.vendordata_address = self.ds_cfg["vendordata_url"]
+
+    @staticmethod
+    def ds_detect():
+        return dmi.read_dmi_data("system-manufacturer") == "Aeza"
+
+    def _get_data(self):
+        system_uuid = dmi.read_dmi_data("system-uuid")
+
+        md = read_metadata(
+            format_url(self.metadata_address, system_uuid),
+            timeout_seconds=self.timeout_seconds,
+            max_wait_seconds=self.max_wait_seconds,
+            sec_between_retries=self.sec_between_retries,
+            retries=self.retries,
+        )
+        ud = read_data(
+            format_url(self.userdata_address, system_uuid),
+            timeout_seconds=self.timeout_seconds,
+            max_wait_seconds=self.max_wait_seconds,
+            sec_between_retries=self.sec_between_retries,
+            retries=self.retries,
+        )
+        vd = read_data(
```

Comment:
This behavior closely aligns with our existing DataSourceNoCloud which reads mandatory metadata and user-data files and optional vendor-data and network-config files. Do you think instead of developing a specialized read_data/read_url set of functions, you could use the existing [util.read_seeded function which will walk throgh md, ud and vd for you with a timeout and retries while coping with missing vendor-data](https://github.com/canonical/cloud-init/blob/main/cloudinit/util.py#L1064)?

Reply:
Fixed in 4e8b32dfd7ce0751ca0781a6894600e2e70f1a2a